pkg/framework/aws_client.go

```diff
@@ -0,0 +1,47 @@
+package framework
+
+import (
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"k8s.io/klog"
+	"k8s.io/utils/ptr"
+)
+
+// AwsClient struct.
+type AwsClient struct {
+	Svc *ec2.EC2
+}
+
+// CreateCapacityReservation Create CapacityReservation.
+func (a *AwsClient) CreateCapacityReservation(instanceType string, instancePlatform string, availabilityZone string, instanceCount int64) (string, error) {
+	input := &ec2.CreateCapacityReservationInput{
+		InstanceType:          aws.String(instanceType),
+		InstancePlatform:      aws.String(instancePlatform),
+		AvailabilityZone:      aws.String(availabilityZone),
+		InstanceCount:         aws.Int64(instanceCount),
+		InstanceMatchCriteria: aws.String("targeted"),
+		EndDateType:           aws.String("unlimited"),
+	}
+	result, err := a.Svc.CreateCapacityReservation(input)
+
+	if err != nil {
+		return "", fmt.Errorf("error creating capacity reservation: %w", err)
+	}
+
+	capacityReservationID := ptr.Deref(result.CapacityReservation.CapacityReservationId, "")
+	klog.Infof("The created capacityReservationID is %s", capacityReservationID)
+
+	return capacityReservationID, err
+}
+
+// CancelCapacityReservation Cancel a CapacityReservation.
+func (a *AwsClient) CancelCapacityReservation(capacityReservationID string) (bool, error) {
+	input := &ec2.CancelCapacityReservationInput{
+		CapacityReservationId: aws.String(capacityReservationID),
+	}
+	result, err := a.Svc.CancelCapacityReservation(input)
+
+	return *result.Return, err
```

Comment:
Here if result is nil, invoking result.Return could yield a panic.
I think we could return *bool and so just do:
```
return a.Svc.CancelCapacityReservation(input)
```

Reply:
Seems `return a.Svc.CancelCapacityReservation(input)` doesn't work, because it doesn't return *bool, it returns *CancelCapacityReservationOutput, there is the definition of CancelCapacityReservation function and CancelCapacityReservationOutput type. I changed this to use ptr.Deref, what do you think?
```
// CancelCapacityReservation API operation for Amazon Elastic Compute Cloud.
//
// Cancels the specified Capacity Reservation, releases the reserved capacity,
// and changes the Capacity Reservation's state to cancelled.
//
// Instances running in the reserved capacity continue running until you stop
// them. Stopped instances that target the Capacity Reservation can no longer
// launch. Modify these instances to either target a different Capacity Reservation,
// launch On-Demand Instance capacity, or run in any open Capacity Reservation
// that has matching attributes and sufficient capacity.
//
// Returns awserr.Error for service API and SDK errors. Use runtime type assertions
// with awserr.Error's Code and Message methods to get detailed information about
// the error.
//
// See the AWS API reference guide for Amazon Elastic Compute Cloud's
// API operation CancelCapacityReservation for usage and error information.
// See also, https://docs.aws.amazon.com/goto/WebAPI/ec2-2016-11-15/CancelCapacityReservation
func (c *EC2) CancelCapacityReservation(input *CancelCapacityReservationInput) (*CancelCapacityReservationOutput, error) {
	req, out := c.CancelCapacityReservationRequest(input)
	return out, req.Send()
}
```

```
type CancelCapacityReservationOutput struct {
	_ struct{} `type:"structure"`

	// Returns true if the request succeeds; otherwise, it returns an error.
	Return *bool `locationName:"return" type:"boolean"`
}
```